doc/gtetrinet.6
.\"                                      Hey, EMACS: -*- nroff -*-
.\" gtetrinet.6 is copyright 2001, 2002 by
.\" Jordi Mallach <jordi@debian.org>
.\"
.\" This is free documentation, see the latest version of the GNU Free
.\" Documentation License for copying conditions. There is NO warranty.
.\"
.TH GTETRINET 6 "October 14, 2020" "gtetrinet"
.\" Please adjust this date whenever revising the manpage.
.\"

.SH NAME
GTetrinet \- a Tetrinet client for X using Gtk+ and GNOME-libs.

.SH SYNOPSIS
.B gtetrinet
.RI [ options ]

.SH DESCRIPTION
This manual page documents briefly the
.B gtetrinet
command.
.PP
\fBGTetrinet\fP is a Tetrinet client using Gtk+ and GNOME-libs.
Tetrinet is a variant of the popular Tetris brick game, that can be played
simultaneously by 6 players. GTetrinet's goal is to remain completely
compatible with the Windows, original client. Tetrinet clients need to
connect to a dedicated server in order to be useful (see USAGE). GTetrinet
only provides the client part.
.SH OPTIONS
GTetrinet accepts the following options, besides the usual Gtk+/GNOME
switches:
.TP
.B \-c, \-\-connect=SERVER
Connect to the specified server.
.TP
.B \-?, \-\-help
Display a summary of commandline options.
.TP
.B \-n, \-\-nickname=NICKNAME
Use the specified nickname on the servers.
.TP
.B \-t, \-\-team=TEAM
Join the specified team name for teamplay.
.TP
.B \-s, \-\-spectate
Connect to the server as an spectator.
.TP
.B \-p, \-\-password=PASSWORD
Use the given password to connect as a spectator (if needed).

.SH USAGE
To play Tetrinet, you first need to find a server where you can connect.
You may try the servers at
.I https://servers.tetrinet.fr/
or create your own using \fBtetrinetx\fP (a UNIX server, see
.I http://tetrinetx.sourceforge.net/
) or \fBJetrix\fP (a Java server, see
.I https://github.com/ebourg/jetrix
for the most recent version).
.PP
Once inside a server, you'll be able to communicate with other players
using the \fBPartyLine\fP tab, see the server's highscores in the
\fBWinlist\fP tab and the actual game in the \fBPlaying Fields\fP tab.
You can switch between the three tabs using the Alt+1, Alt+2 and Alt+3
shortcuts.
.br
Tetrinet is like Tetris, but multiplayer and with some additions. The goal of
the game is to remain alive, by filling up other player's fields. There are
some special blocks that can be used for your own benefit and against other
players. They will appear randomly in your field, replacing previously
existing blocks when you complete a line. You'll gain special blocks when
you complete a line that contains one. It will be added to your
\fBinventory\fP, which is just below your playfield. You need to use them in
the order you collect them, by pressing the number of the player you want to
use it against. For example, if your active block is "Add Line" and you want
to use it against Player 3, you'd have to press "3" and Player 3 would
get a line added in that instant.
The following blocks are available:
.RS
.TP
.I A block
Add Line will add an incomplete line to the bottom of the selected player's
field.
.TP
.I B block
This will remove all the special blocks from a player's field.
.TP
.I C block
Clear Line is the opposite to Add Line; it will remove a line from the
top of the selected player's field.
.TP
.I G block
The Gravity block may be very useful if applied to a field which has lots of
unfilled spaces all around. Blocks over them will fall into the gaps,
making the puzzle collapse significantly.
.TP
.I N block
Nuke will clear the chosen playing field completely, as if they had just
started the game. It's useful to help team mates or when you are about to
lose.
.TP
.I O block
Block Bomb will clear 3x3 portions of a player's field where there are other
O blocks. The blocks in the 3x3 area will get scattered anywhere around the
field.
.TP
.I Q block
Quake will make each of the lines in a player's field shift to the left or
right, or stay as they were.
.TP
.I R block
Random will remove 10 random blocks from the selected player's field.
.TP
.I S block
Switch allows you to swap your playing field with another player. This can
be a deadly block if used with a bit of strategy.
.RE

.SH HOMEPAGE
http://gtetrinet.sourceforge.net/

.SH AUTHOR
This manual page was written by Jordi Mallach <jordi@debian.org>,
for the Debian system (but may be used by others).
Ka-shu Wong <zillidot@zillidot.net> is GTetrinet's main developer.
You can find more information about GTetrinet in /usr/share/doc/gtetrinet.


NEWS
GTetrinet 0.7.11 - 2006-11-15
- Fix a long-standing hang/crash when executing GTetrinet for
  the first time (Loïc Minier, #114425).
- Remember game mode setting (Lucas Nussbaum, #344522).
- Fix compilation without libesd (Ricardo Setti, #312995).
- Fix a crash on NetBSD/Sparc64 (#338712).
- New and updated translations: Nepali (Pawan Chitrakar),
  Catalan (Jordi Mallach), Welsh (Dafydd Harries), Dutch (Wouter Bolsterlee),
  Swedish (Daniel Nylander).

GTetrinet 0.7.10 - 2006-09-02
- SECURITY RELEASE -- Ref. CVE-2006-3125
- Add index undeflow protections in network code.
- Require GTK+ 2.6.0.
- Fixed manpage typos.
- Require new intltool and move ALL_LINGUAS to po/LINGUAS.
- New and updated translations: Bulgarian (Alexander Shopov),
  Basque (Iñaki Larrañaga), Traditional Chinese (Abel Cheung),
  German (Jens Seidel), Nepali (Pawal Chitrakar), Finnish (Ilkka Tuohela),
  Vietnamese (Clytie Siddall).

GTetrinet 0.7.9 - 2005-04-29
- Revamped encoding of data handling, fixing the output in partyline, etc.
  when using UTF-8 locales (Vidar Holen).
- Ported to use GtkAboutDialog (Pedro Villavicencio Garrido).
- Fixed network latency issues (Julien Plissonneau Duquène).
- Fixed crashes in the channel list support (Julien Plissonneau Duquène).
- New and updated translations: Traditional Chinese (GNOME HK Team),
  Canadian English (Adam Weinberger), Norwegian bokmål (Vidar Holen),
  Bulgarian (Yavor Doganov), Kinyarwanda (Steve Murphy).

GTetrinet 0.7.8 - 2004-12-26
- Added minimal support for the Blocktrix protocol, now in use at
  tetrinet.org (Loren Abrams).
- New and updated translations: Italian (Riccardo Bozzo),
  German (Gerfried Fuchs), Korean (Michael Kim),
  Canadian English (Adam Weinberger), Serbian (Danilo Šegan),
  Punjabi (Amanpreet Singh Alam), Norwegian bokmål (Kjartan Maraas),
  Simplified Chinese (Funda Wang).

GTetrinet 0.7.7 - 2004-04-18
- Regenerated tarball using automake 1.4 to workaround
  tar --posix problems.

GTetrinet 0.7.6 - 2004-04-16
- Added a button in the Preferences dialog that opens a browser window
  with the GTetrinet themes webpage. (Dani)
- Made the main window resizable. It doesn't mean that the fields will
  be resizable too ;) (Dani)
- Now you can rotate and move a piece while holding CTRL. This fixes the
  bug reported by Christian Kröner. (Dani)
- Disable deprecated flags which broke the build with
  GTK 2.4 (Kjartan Maraas).
- New and updated translations: Norwegian (Kjartan Maraas),
  Croatian (Robert Sedak), Irish (Alastair McKinstry),
  Brazilian Portuguese (Steve Lacerda and Estêvão Samuel Procópio),
  British English (Gareth Owen), Canadian English (Adam Weinberger),
  Welsh (Dafydd Harries), Catalan (Jordi Mallach), Swedish (Christian Rose),
  Czech (Miloslav Trmac), Portuguese (Duarte Loreto),
  Dutch (Vincent van Adrighem), Polish (Artur Flinta),
  French (Christophe Merlet), Spanish (Dani Carbonell).

GTetrinet 0.7.5 - 2003-10-04
- HIGification improvements in several dialogs (Dagfinn Ilmari Mannsåker).
- Make the default keys usable in azerty keyboards (Dani).
- New and updated translations: Azerbaijani (Metin Amiroff),
  Polish (Artur Flinta), Japanese (KAMAGASAKO Masatoshi).

GTetrinet 0.7.4 - 2003-08-29
- Fix configure.in so nobody needs intl/ dir. (Jordi)
- Connect and Team dialogs now have "OK" as default action (Dagfinn Ilmari
  Mannsåker).
- Fix a crash when restoring the default keys (Dani).
- Cursor fixes for the game fields (Dani).
- Fix bug where keyboard would become unusable on game start (Benjamin Drieu).
- Disable channel list by default to avoid problems with weird servers (Dani).
- New and updated translations: Spanish (Francisco Javier F. Serrador),
  Czech (Miloslav Trmac), Dutch (Kees van den Broek), Polish (Artur Flinta),
  Serbian (Danilo Segan), Malay (Hasbullah Bin Pit), Catalan (Jordi Mallach),
  Welsh (Dafydd Harries), Traditional Chinese (Abel Cheung),
  German (Gerfried Fuchs).

GTetrinet 0.7.3 - 2003-06-23
- Fixed search path for gtetrinet.png broken in 0.7.2.
- Merged Connect/Disconnect buttons.
- Fixed annoying bug in keyboard handling code which could make
  GTetrinet beep on each keypress.
- New and updated translations: Traditional Chinese (Abel Cheung),
  Dutch (Tino Meinen), Macedonian (Arangel Angov).

GTetrinet 0.7.2 - 2003-06-07
- Added new icons to several places.
- Merged Start/Stop button.
- Nickname autocompletion in partyline!
- Show /msg commands in partyline.
- The mouse cursor is now hidden while playing.
- New keyboard shortcuts for connect/disconnect, change team and
  start/stop/pause game.
­ Attack/defense keys are now configurable, helping Azerty keyboard users.
- Other bugfixes.
- New and updated translations: French (Christophe Merlet), Catalan
  (Jordi Mallach), Swedish (Christian Rose), Portuguese (Duarte Loreto),
  Czech (Miloslav Trmac), Dutch (Tino Meinen), Brazilian Portuguese
  (Alexandre Folle de Menezes), Japanese (KAMAGASAKO Masatoshi), Serbian
  (Danilo Segan), Basque (Julen Landa Alustiza), Traditional Chinese
  (Abel Cheung).

GTetrinet 0.7.1 - 2003-04-14
- Added an option to disable the channels list feature, so GTetrinet is
  usable in non-supported (yet) servers.
- Added channel list support for two new server types.
- Fixed segfault in the partyline when entering Unicode control characters
  in the history.
- Now installs .desktop file in the GNOME2 location.
- New and updated translations: Spanish (Francisco Javier Fernández),
  Japanese (KAMAGASAKO Masatoshi), Catalan (Jordi Mallach), Swedish
  (Christian Rose), German (Gerfried Fuchs), Portuguese (Duarte Loreto),
  Czech (Miloslav Trmac).

GTetrinet 0.7.0 - 2003-03-17
- Added a list of available channels in partyline, and the ability to
  join them by double-clicking on them.
- Added timestamps to partyline messages.
- Optimizations and bugfixes.
- New and updated translations: Swedish (Christian Rose), Catalan
  (Jordi Mallach), Dutch (Tino Meinen), Portuguese (Duarte Loreto),

src/gtetrinet.c

/*
 *  GTetrinet
 *  Copyright (C) 1999, 2000, 2001, 2002, 2003  Ka-shu Wong (kswong@zip.com.au)
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <libintl.h>
#include <gtk/gtk.h>
#include <stdlib.h>
#include <time.h>
#include <sys/poll.h>
#include <sys/types.h>
#include <gobject/gtype.h>
#include <signal.h>
#include <popt.h>

#include "gtetrinet.h"
#include "gtet_config.h"
#include "client.h"
#include "tetrinet.h"
#include "tetris.h"
#include "fields.h"
#include "partyline.h"
#include "winlist.h"
#include "misc.h"
#include "commands.h"
#include "sound.h"
#include "string.h"

#include "images/fields.xpm"
#include "images/partyline.xpm"
#include "images/winlist.xpm"

static GtkWidget *pixmapdata_label (char **d, char *str);
static int gtetrinet_key (int keyval, int mod);
gint keypress (GtkWidget *widget, GdkEventKey *key);
gint keyrelease (GtkWidget *widget, GdkEventKey *key);
void switch_focus (GtkNotebook *notebook,
                   void *page,
                   guint page_num);

static GtkWidget *pfields, *pparty, *pwinlist;
static GtkWidget *winlistwidget, *partywidget, *fieldswidget;
static GtkWidget *notebook;

GtkWidget *app;

char *option_connect = 0, *option_nick = 0, *option_team = 0, *option_pass = 0;
int option_spec = 0;

int gamemode = ORIGINAL;

int fields_width, fields_height;

gulong keypress_signal;

GSettings* settings;
GSettings* settings_keys;
GSettings* settings_themes;

static const struct poptOption options[] = {
    {"connect", 'c', POPT_ARG_STRING, &option_connect, 0, ("Connect to server"), ("SERVER")},
    {"nickname", 'n', POPT_ARG_STRING, &option_nick, 0, ("Set nickname to use"), ("NICKNAME")},
    {"team", 't', POPT_ARG_STRING, &option_team, 0, ("Set team name"), ("TEAM")},
    {"spectate", 's', POPT_ARG_NONE, &option_spec, 0, ("Connect as a spectator"), NULL},
    {"password", 'p', POPT_ARG_STRING, &option_pass, 0, ("Spectator password"), ("PASSWORD")},
    {NULL, 0, 0, NULL, 0, NULL, NULL}
};

static int gtetrinet_poll_func(GPollFD *passed_fds,
                               guint nfds,
                               int timeout)
{ /* passing a timeout wastes time, even if data is ready... don't do that */
  int ret = 0;
  struct pollfd *fds = (struct pollfd *)passed_fds;

  ret = poll(fds, nfds, 0);
  if (!ret && timeout)
    ret = poll(fds, nfds, timeout);

  return (ret);
}

/*
 * based on https://developer.gnome.org/gio/stable/gio-GSettingsSchema-GSettingsSchemaSource.html
 * I have no idea why this is not the default behavior
 */
GSettings *get_schema_settings(const gchar *schema_id)
{
  GSettingsSchema *schema;
  GSettingsSchemaSource *schema_source;
  schema_source = g_settings_schema_source_new_from_directory(GSETTINGSSCHEMADIR, g_settings_schema_source_get_default(), FALSE, NULL); // @suppress("Symbol is not resolved")
  schema = g_settings_schema_source_lookup(schema_source, schema_id, FALSE);
  if (schema == NULL)
  {
    return g_settings_new(schema_id);
  }
  return g_settings_new_full(schema, NULL, NULL);
}

int main (int argc, char *argv[])
{
    GtkWidget *label;
    GdkPixbuf *icon_pixbuf;
    GError *err = NULL;
    
    bindtextdomain(PACKAGE, LOCALEDIR);
    bind_textdomain_codeset(PACKAGE, "UTF-8");
    textdomain(PACKAGE);

    srand (time(NULL));

    /*
    gnome_program_init (APPID, APPVERSION, LIBGNOMEUI_MODULE,
                        argc, argv, GNOME_PARAM_POPT_TABLE, options,
                        GNOME_PARAM_NONE);
    */
    GOptionEntry options[] = { {NULL}};
    if (!gtk_init_with_args(&argc,&argv,"gtetrinet",options,NULL,&err))
    {
        fprintf (stderr, "Failed to init GTK: %s\n", err->message);
        g_error_free(err);
        err = NULL;
        return 1;
    }
    textbox_setup (); /* needs to be done before text boxes are created */

    // First, try to get settings from compiled schema directory (as we can properly check these), then try generic system directories chosen by gsettings library
    settings = get_schema_settings (GSETTINGS_DOMAIN);
    settings_keys = get_schema_settings (GSETTINGS_DOMAIN_KEYS);
    settings_themes = get_schema_settings (GSETTINGS_DOMAIN_THEMES);

    g_signal_connect_swapped (settings, "changed", G_CALLBACK(config_loadconfig), NULL);
    g_signal_connect_swapped (settings_keys, "changed", G_CALLBACK(config_loadconfig_keys), NULL);
    g_signal_connect_swapped (settings_themes, "changed", G_CALLBACK(config_loadconfig_themes), NULL);

    /* load settings */
    config_loadconfig ();
    config_loadconfig_keys ();

    /* first set up the display */

    /* create the main window */
    app = gtk_window_new (GTK_WINDOW_TOPLEVEL);
    gtk_window_set_title (GTK_WINDOW (app), APPNAME);

    g_signal_connect (G_OBJECT(app), "destroy",
                        G_CALLBACK(destroymain), NULL);
    keypress_signal = g_signal_connect (G_OBJECT(app), "key-press-event",
                                        G_CALLBACK(keypress), NULL);
    g_signal_connect (G_OBJECT(app), "key-release-event",
                        G_CALLBACK(keyrelease), NULL);
    gtk_widget_set_events (app, GDK_KEY_PRESS_MASK | GDK_KEY_RELEASE_MASK);

    gtk_window_set_resizable (GTK_WINDOW (app), TRUE);
    
    /* create and set the window icon */
    icon_pixbuf = gdk_pixbuf_new_from_file (PIXMAPSDIR "/gtetrinet.png", NULL);
    if (icon_pixbuf)
    {
      gtk_window_set_icon (GTK_WINDOW (app), icon_pixbuf);
      g_object_unref (icon_pixbuf);
    }

    /* create the notebook */
    notebook = gtk_notebook_new ();
    gtk_notebook_set_tab_pos (GTK_NOTEBOOK(notebook), GTK_POS_TOP);

    /* put it in the main window */
    gtk_container_add (GTK_CONTAINER(app), notebook);

    /* make menus + toolbar */
    make_menus (GTK_WINDOW(app));

    /* create the pages in the notebook */
    fieldswidget = fields_page_new ();
    gtk_widget_set_sensitive (fieldswidget, TRUE);
    gtk_widget_show (fieldswidget);
    pfields = gtk_box_new (GTK_ORIENTATION_HORIZONTAL, 0);
    gtk_container_set_border_width (GTK_CONTAINER(pfields), 0);
    gtk_container_add (GTK_CONTAINER(pfields), fieldswidget);
    gtk_widget_show (pfields);
    g_object_set_data (G_OBJECT(fieldswidget), "title", "Playing Fields"); // FIXME
    label = pixmapdata_label (fields_xpm, "Playing Fields");
    gtk_widget_show (label);
    gtk_notebook_append_page (GTK_NOTEBOOK(notebook), pfields, label);

    partywidget = partyline_page_new ();
    gtk_widget_show (partywidget);
    pparty = gtk_box_new (GTK_ORIENTATION_HORIZONTAL, 0);
    gtk_container_set_border_width (GTK_CONTAINER(pparty), 0);
    gtk_container_add (GTK_CONTAINER(pparty), partywidget);
    gtk_widget_show (pparty);
    g_object_set_data (G_OBJECT(partywidget), "title", "Partyline"); // FIXME
    label = pixmapdata_label (partyline_xpm, "Partyline");
    gtk_widget_show (label);
    gtk_notebook_append_page (GTK_NOTEBOOK(notebook), pparty, label);

    winlistwidget = winlist_page_new ();
    gtk_widget_show (winlistwidget);
    pwinlist = gtk_box_new (GTK_ORIENTATION_HORIZONTAL, 0);
    gtk_container_set_border_width (GTK_CONTAINER(pwinlist), 0);
    gtk_container_add (GTK_CONTAINER(pwinlist), winlistwidget);
    gtk_widget_show (pwinlist);
    g_object_set_data (G_OBJECT(winlistwidget), "title", "Winlist"); // FIXME
    label = pixmapdata_label (winlist_xpm, "Winlist");
    gtk_widget_show (label);
    gtk_notebook_append_page (GTK_NOTEBOOK(notebook), pwinlist, label);

    /* add signal to focus the text entry when switching to the partyline page*/
    g_signal_connect_after(G_OBJECT (notebook), "switch-page",
		           G_CALLBACK (switch_focus),
		           NULL);

    gtk_widget_show (notebook);
    g_object_set (G_OBJECT (notebook), "can-focus", FALSE, NULL);

    partyline_show_channel_list (list_enabled);
    gtk_widget_show (app);

//    gtk_widget_set_size_request (partywidget, 480, 360);
//    gtk_widget_set_size_request (winlistwidget, 480, 360);

    /* initialise some stuff */
    config_loadconfig_themes ();
    commands_checkstate ();

    /* check command line params */
#ifdef DEBUG
    printf ("option_connect: %s\n"
            "option_nick: %s\n"
            "option_team: %s\n"
            "option_pass: %s\n"
            "option_spec: %i\n",
            option_connect, option_nick, option_team,
            option_pass, option_spec);
#endif
    if (option_nick) GTET_O_STRCPY(nick, option_nick);
    if (option_team) GTET_O_STRCPY(team, option_team);
    if (option_pass) GTET_O_STRCPY(specpassword, option_pass);
    if (option_spec) spectating = TRUE;
    if (option_connect) {
        client_init (option_connect, nick);
    }

    /* Don't schedule if data is ready, glib should do this itself,
     * but welcome to anything that works... */
    g_main_context_set_poll_func(NULL, gtetrinet_poll_func);

    /* gtk_main() */
    gtk_main ();

    g_object_unref (settings);
    g_object_unref (settings_keys);
    g_object_unref (settings_themes);

    client_disconnect ();
    /* cleanup */
    fields_cleanup ();

    return 0;
}

GtkWidget *pixmapdata_label (char **d, char *str)
{
    GdkPixbuf *pb;
    GtkWidget *box, *widget;

    box = gtk_box_new (GTK_ORIENTATION_HORIZONTAL, 0);

    pb = gdk_pixbuf_new_from_xpm_data ((const char **)d);
    widget = gtk_image_new_from_pixbuf (pb);
    gtk_widget_show (widget);
    gtk_box_pack_start (GTK_BOX(box), widget, TRUE, TRUE, 0);
  
    widget = gtk_label_new (str);
    gtk_widget_show (widget);
    gtk_box_pack_start (GTK_BOX(box), widget, TRUE, TRUE, 0);

    return box;
}

/* called when the main window is destroyed */
void destroymain (void)
{
    gtk_main_quit ();
}

/*
 The key press/release handlers requires a little hack:
 There is no indication whether each keypress/release is a real press
 or a real release, or whether it is just typematic action.
 However, if it is a result of typematic action, the keyrelease and the
 following keypress event have the same value in the time field of the
 GdkEventKey struct.
 The solution is: when a keyrelease event is received, the event is stored
 and a timeout handler is installed.  if a subsequent keypress event is
 received with the same value in the time field, the keyrelease event is
 discarded.  The keyrelease event is sent if the timeout is reached without
 being cancelled.
 This results in slightly slower responses for key releases, but it should not
 be a big problem.
 */

GdkEventKey k;
gint keytimeoutid = 0;

gint keytimeout (gpointer data)
{
    tetrinet_upkey (k.keyval);
    keytimeoutid = 0;
    return FALSE;
}

gint keypress (GtkWidget *widget, GdkEventKey *key)
{
    int game_area;

    if (widget == app)
    {
      int cur_page = gtk_notebook_get_current_page(GTK_NOTEBOOK(notebook));
      int pfields_page = gtk_notebook_page_num(GTK_NOTEBOOK(notebook),
                                               pfields);
      /* Main window - check the notebook */
      game_area = (cur_page == pfields_page);
    }
    else
    {
        /* Sub-window - find out which */
        char *title = NULL;

        title = g_object_get_data(G_OBJECT(widget), "title");
        game_area =  title && !strcmp( title, "Playing Fields");
    }

    if (game_area)
    { /* keys for the playing field - key releases needed - install timeout */
      if (keytimeoutid && key->time == k.time)
        g_source_remove (keytimeoutid);
    }

    /* Check if it's a GTetrinet key */
    if (gtetrinet_key (key->keyval, key->state & (GDK_MOD1_MASK)))
    {
      g_signal_stop_emission_by_name (G_OBJECT(widget), "key-press-event");
      return TRUE;
    }

/*    if ((key->state & (GDK_MOD1_MASK | GDK_CONTROL_MASK)) > 0)
    return FALSE;*/
    
    if (game_area && ingame && (gdk_keyval_to_lower (key->keyval) == keys[K_GAMEMSG]))
    {
      g_signal_handler_block (app, keypress_signal);
      fields_gmsginputactivate (TRUE);
      g_signal_stop_emission_by_name (G_OBJECT(widget), "key-press-event");
    }

    if (game_area && tetrinet_key (key->keyval))
    {
      g_signal_stop_emission_by_name (G_OBJECT(widget), "key-press-event");
      return TRUE;
    }
    
    return FALSE;
}

gint keyrelease (GtkWidget *widget, GdkEventKey *key)
{
    int game_area;

    if (widget == app)
    {
      int cur_page = gtk_notebook_get_current_page(GTK_NOTEBOOK(notebook));
      int pfields_page = gtk_notebook_page_num(GTK_NOTEBOOK(notebook),
                                               pfields);
      /* Main window - check the notebook */
      game_area = (cur_page == pfields_page);
    }
    else
    {
        /* Sub-window - find out which */
        char *title = NULL;

        title = g_object_get_data(G_OBJECT(widget), "title");
        game_area =  title && !strcmp( title, "Playing Fields");
    }

    if (game_area)
    {
        k = *key;
        keytimeoutid = g_timeout_add (10, keytimeout, 0);
        g_signal_stop_emission_by_name (G_OBJECT(widget), "key-release-event");
        return TRUE;
    }
    return FALSE;
}

/*
 TODO: make this switch between detached pages too
 */
static int gtetrinet_key (int keyval, int mod)
{
  if (mod != GDK_MOD1_MASK)
    return FALSE;
    
  switch (keyval)
  {
  case GDK_KEY_1: gtk_notebook_set_current_page (GTK_NOTEBOOK(notebook), 0); break;
  case GDK_KEY_2: gtk_notebook_set_current_page (GTK_NOTEBOOK(notebook), 1); break;
  case GDK_KEY_3: gtk_notebook_set_current_page (GTK_NOTEBOOK(notebook), 2); break;
  default:
    return FALSE;
  }
  return TRUE;
}

/* funky page detach stuff */

/* Type to hold primary widget and its label in the notebook page */
typedef struct {
    GtkWidget *parent;
    GtkWidget *widget;
    int pageNo;
} WidgetPageData;

void destroy_page_window (GtkWidget *window, gpointer data)
{
    WidgetPageData *pageData = (WidgetPageData *)data;

    /* Put widget back into a page */
    //gtk_widget_reparent (pageData->widget, pageData->parent);
    g_object_ref (pageData->parent);
    gtk_container_remove(GTK_CONTAINER (gtk_widget_get_parent (pageData->widget)), pageData->widget);
    gtk_container_add(GTK_CONTAINER (pageData->parent), pageData->widget);
    g_object_unref (pageData->parent);

    /* Select it */
    gtk_notebook_set_current_page (GTK_NOTEBOOK(notebook), pageData->pageNo);

    /* Free return data */
    g_free (data);
}

void move_current_page_to_window (void)
{
    WidgetPageData *pageData;
    GtkWidget *page, *child, *newWindow;
    GList *dlist;
    gint pageNo;
    char *title;

    /* Extract current page's widget & it's parent from the notebook */
    pageNo = gtk_notebook_get_current_page (GTK_NOTEBOOK(notebook));
    page   = gtk_notebook_get_nth_page (GTK_NOTEBOOK(notebook), pageNo );
    dlist  = gtk_container_get_children (GTK_CONTAINER(page));
    if (!dlist ||  !(dlist->data))
    {
        /* Must already be a window */
        if (dlist)
           g_list_free (dlist);
        return;
    }
    child = (GtkWidget *)dlist->data;
    g_list_free (dlist);

    /* Create new window for widget, plus container, etc. */
    newWindow = gtk_window_new (GTK_WINDOW_TOPLEVEL);
    title = g_object_get_data (G_OBJECT(child), "title");
    if (!title)
        title = "GTetrinet";
    gtk_window_set_title (GTK_WINDOW (newWindow), title);
    gtk_container_set_border_width (GTK_CONTAINER (newWindow), 0);

    /* Attach key events to window */
    g_signal_connect (G_OBJECT(newWindow), "key-press-event",
                        G_CALLBACK(keypress), NULL);
    g_signal_connect (G_OBJECT(newWindow), "key-release-event",
                        G_CALLBACK(keyrelease), NULL);
    gtk_widget_set_events (newWindow, GDK_KEY_PRESS_MASK | GDK_KEY_RELEASE_MASK);
    gtk_window_set_resizable (GTK_WINDOW(newWindow), TRUE);

    /* Create store to point us back to page for later */
    pageData = g_new( WidgetPageData, 1 );
    pageData->parent = page;
    pageData->widget = child;
    pageData->pageNo = pageNo;

    /* Move main widget to window */
    //gtk_widget_reparent (child, newWindow);
    g_object_ref (child);
    gtk_container_remove(GTK_CONTAINER (gtk_widget_get_parent (child)), child);
    gtk_container_add(GTK_CONTAINER (newWindow), child);
    g_object_unref (child);


    /* Pass ID of parent (to put widget back) to window's destroy */
    g_signal_connect (G_OBJECT(newWindow), "destroy",
                        G_CALLBACK(destroy_page_window),
                        (gpointer)(pageData));

    gtk_widget_show_all( newWindow );

    /* cure annoying side effect */
    if (gmsgstate)
        fields_gmsginput(TRUE);
    else
        fields_gmsginput(FALSE);

}

/* show the fields notebook tab */
void show_fields_page (void)
{
    gtk_notebook_set_current_page (GTK_NOTEBOOK(notebook), 0);
}

/* show the partyline notebook tab */
void show_partyline_page (void)
{
    gtk_notebook_set_current_page (GTK_NOTEBOOK(notebook), 1);
}

void unblock_keyboard_signal (void)
{
    g_signal_handler_unblock (app, keypress_signal);
}

void switch_focus (GtkNotebook *notebook,
                   void *page,
                   guint page_num)
{
    if (connected)
      switch (page_num)
      {
        case 0:
          if (gmsgstate) fields_gmsginputactivate (1);
          else partyline_entryfocus ();
          break;
        case 1: partyline_entryfocus (); break;
        case 2: winlist_focus (); break;
      }
}
